app/liquid_tags/liquid_tag_base.rb

```diff
@@ -3,6 +3,12 @@ def self.script
     ""
   end
 
+  def detect_in_context(key, context)
+    # Pulls the value of the specified key out of context.environments which is
+    # an Array of Hashes. Otherwise, it returns nil.
+    context.environments.detect { |c| c.key? key }.try(:[], key)
+  end
+
```

Comment:
Said helper [mentioned above](https://github.com/thepracticaldev/dev.to/pull/8871/files#r445053974).

`context` will look something like:
```ruby
[{:source => Article}, {}]
```
So my approach is trying to "future proof" it a little bit.

Reply:
Rather than looping through the environments why not just set the key in the context hash? It seems like that would be a lot more accessible according to the docs which show an example of fetching a key from the context.
```ruby
def render(context)
  (@factor * context["multiply_by"].to_f).to_s
end
```

Reply:
Agreed - for the life of me, I couldn't get that to work. I tried both ways of setting an assignment:
```ruby
@template = Liquid::Template.parse("{% multiply 5 %}")
@template.assigns["multiply_by"] = 10
@template.render    # => "50"
```
and the current way
```ruby
@template = Liquid::Template.parse("{% multiply 5 %}")
@template.render('multiply_by' => 3)    # => 15
```

Both ways and playing with it, the argument still always shows up in the `context.environments` Array 🤔 . I eventually chalked it up as stale docs since those were last updated in 2016 😬 .

Reply:
This has me thinking though...I've been playing around in `pry` and the source code for the gem. I'll update this. At a minimum, I shouldn't be passing it an Object as that's not a "permitted type". And `context` has its own built-in `find_variable` method, but won't work with an Object 🤔 .

Reply:
@mstruve thanks for keeping me honest with this inquiry 💯 .

**TLDR** - I [switched to register](https://github.com/Shopify/liquid/wiki/Liquid-for-Programmers#difference-between-assigns-and-registers) and removed the looping through environments and those docs/that wiki are definitely stale.

**Details** _for future reference_
`render` [will create a new `Context` object](https://github.com/Shopify/liquid/blob/1feaa6381300d56e2c71b49ad8fee0d4b625147b/lib/liquid/template.rb#L189) which will [`squash_instance_assigns_with_environments`](https://github.com/Shopify/liquid/blob/1feaa6381300d56e2c71b49ad8fee0d4b625147b/lib/liquid/context.rb#L37). So even if you create your own `Context` object to pass to `render`, inside `Context` if you try to use `environments`, `scopes`, or `instance_assigns` they will end up in the `environments` or `scopes` Array. Neither of those is really appropriate here. Instead, I switched to register which sounds more appropriate.

Registers [are set](https://github.com/Shopify/liquid/blob/1feaa6381300d56e2c71b49ad8fee0d4b625147b/lib/liquid/context.rb#L31) instead of the whole Array riff-raff. Also, a [code comment](https://github.com/Shopify/liquid/blob/1feaa6381300d56e2c71b49ad8fee0d4b625147b/lib/liquid/template.rb#L170-L171) points out that registers are
> hash with register variables. Those can be accessed from filters and tags and might be useful to integrate liquid more with its host application.

🎉